ml-models/data_utils.py

```diff
@@ -0,0 +1,68 @@
+import numpy as np
+import pandas as pd
+from sklearn.preprocessing import StandardScaler
+from sklego.preprocessing import RepeatingBasisFunction
+from tqdm import tqdm
+
+
+def get_processed_data(path):
+    # Remove duplicate readings from the same hour
+    # Occurs when different weather enums are combined, but all other values are the same
+    cols = ['dt', 'dt_iso', 'temp', 'dew_point', 'pressure', 'humidity', 'rain_1h', 'snow_1h', 'weather_main']
+    df = pd.read_csv(path, usecols=cols).drop_duplicates(['dt'])
+    dates = pd.to_datetime(df['dt_iso'], format='%Y-%m-%d %H:%M:%S +0000 UTC')
+
+    # Fill NaN values with 0
+    df = df.fillna(0)
+
+    # One-hot encoding of weather_main enum
+    df = pd.concat([df, pd.get_dummies(df['weather_main'])], axis=1)
+    df = df.drop(columns=['weather_main', 'Clear', 'Clouds', 'Drizzle', 'Dust', 'Fog', 'Haze', 'Mist', 'Thunderstorm',
+                          'Tornado', 'dt', 'dt_iso'])
+
+    # Sine / cosine encoding of the hour of day
+    # sec_in_day = 60 * 60 * 24
+    # df['day_sin'] = np.sin(df['dt'].to_numpy() * (2 * np.pi / sec_in_day))
+    # df['day_cos'] = np.cos(df['dt'].to_numpy() * (2 * np.pi / sec_in_day))
+    # Radial basis functions might be better
+
+    # Radial basis functions
+    # (https://developer.nvidia.com/blog/three-approaches-to-encoding-time-information-as-features-for-ml-models/)
+    day_in_year = dates.dt.dayofyear.to_numpy()
+    rbf = RepeatingBasisFunction(n_periods=12, input_range=(1, 365))
+    day_in_year_df = pd.DataFrame(data=day_in_year)
+    rbf_df = pd.DataFrame(data=rbf.fit_transform(day_in_year_df),
+                          index=df.index,
+                          columns=['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec'])
+    df = pd.concat([df, rbf_df], axis=1)
+
+    hour_in_day = dates.dt.hour.to_numpy()
+    rbf2 = RepeatingBasisFunction(n_periods=12, input_range=(0, 23))
```

Comment:
Lidt mere sigende navn måske, går ud fra rbf er for ét år, mens rbf2 er for timerne i en dag

Reply:
specifikt linje 40. jeg måtte ikke ændre min selection.

Reply:
True, har ændret `rbf` til `rbf_months` og `rbf2` til `rbf_hours`